platform/src/ActivityValidator.js

```diff
@@ -0,0 +1,208 @@
+import { ConfigValidationError } from './ConfigValidationError';
+
+const ERROR_CATEGORY = "ERROR_CONFIG_DEF"
+
+const errorFileType =  {
+    ACTIVITY: "ActivityConfig",
+    TOOL: "ToolConfig"
+}
+
+class ActivityValidator {
+
+    // Activity checks
+    
+    /**
+     * Check layout panel ids exist
+     * @param {Object} activity - An activity with all references resolved.
+     * @returns {ConfigValidationError[]} Array of configuration errors, empty if there are none.
+     */
+    static checkLayoutPanelIdsExist(activity){
+        let errors = [];
+
+        let combinedRows = [];
+        activity.layout.area.forEach( (row) => combinedRows= combinedRows.concat(row) );
+
+        const ids = new Set(combinedRows);
+        ids.delete(""); // Remove empty references used to indicate extend panel
+
+        ids.forEach( (id) => {
+            if (!this.idExists(activity.panels, id)){
+                    errors.push( 
+                        new ConfigValidationError(ERROR_CATEGORY, "A panel does not exist for the given id.",  
+                        `${activity.id} -> layout -> ${id}`, errorFileType.ACTIVITY)
+                );
+            }
+        });
+
+        return errors;
+    }
+
+    /**
+     * Check the actions reference ids
+     * @param {*} activity - An activity with all references resolved.
+     * @returns {ConfigValidationError[]} Array of configuration errors, empty if there are none.
+     */
+    static checkActions(activity){
+        let errors = [];
+
+        activity.actions.forEach( (act) => {
+            errors = errors.concat( this.checkActionIdsExist(act, activity.panels) );
+        });
+
+        return errors;
+    }
+
+    /**
+     * Check an action panel ids exist
+     * @param {*} action - The action to check
+     * @param {*} panels - The panels 
+     * @returns {ConfigValidationError[]} Array of configuration errors, empty if there are none.
+     */
+    static checkActionIdsExist(action, panels){
+        let errors = [];
+
+        if (!this.idExists( panels, action.source.id)){ // action ids are resolved so get id from object
+            errors.push( 
+                new ConfigValidationError(ERROR_CATEGORY, "A panel does not exist for the given id.",  
+                `actions -> source: ${action.source}`, errorFileType.ACTIVITY)
+            );
```

Comment:
Not `${action.source.id}`?

Reply:
And how does the error message identify the specific action?

Reply:
Hmm, but on Line 64 just a little bit above, you are using `action.source.id`. Wouldn't this be undefined in that place in that case?

Reply:
Yes the idExists() function that's called on line 64 returns false if an null/undefined id is given, maybe this obscures the validation condition. The null check could be made the condition and a comment to note why to make it clearer.

Reply:
The null check of the resolved panel id e.g. `action.source.id` is all that is needed to determine whether the panels id exists for the `checkActionIdsExist()` checks because `action.source` etc have been resolved earlier from a panel id and replaced with the panel object. So an alternative check that is equivalent is checking whether `action.source` is an object. 

The idea behind using `idExists()` is to double-check that the id of a resolved panel exists within the activities' panels for robustness.

Reply:
I would suggest making this explicit by doing both checks: `action.source isObject && checkActionIdsExist(...`. WDYT?

Reply:
Yes I agree and will update.

Reply:
Fixed in 40616a5525ed9b6fb0d7b2dc4de5fc691599691c